File: Shared/RepoPressCoreContracts/dotnet/src/RepoPress.Core/PublishConflictDiffBuilder.cs
using System;
using System.Collections.Generic;

namespace RepoPress.Core
{
    public enum PublishConflictDiffLineKind
    {
        Same,
        Remote,
        Local
    }

    public sealed class PublishConflictDiffLine
    {
        public PublishConflictDiffLine(
            int id,
            PublishConflictDiffLineKind kind,
            string text)
        {
            if (id < 0)
            {
                throw new ArgumentOutOfRangeException(nameof(id));
            }

            Id = id;
            Kind = kind;
            Text = text ?? throw new ArgumentNullException(nameof(text));
        }

        public int Id { get; }

        public PublishConflictDiffLineKind Kind { get; }

        public string Text { get; }

        public string Marker
        {
            get
            {
                switch (Kind)
                {
                    case PublishConflictDiffLineKind.Same:
                        return " ";
                    case PublishConflictDiffLineKind.Remote:
                        return "-";
                    case PublishConflictDiffLineKind.Local:
                        return "+";
                    default:
                        throw new ArgumentOutOfRangeException();
                }
            }
        }
    }

    public sealed class PublishConflictDiffBuilder
    {
        public IReadOnlyList<PublishConflictDiffLine> Diff(string remote, string local)
        {
            if (remote == null)
            {
                throw new ArgumentNullException(nameof(remote));
            }

            if (local == null)
            {
                throw new ArgumentNullException(nameof(local));
            }

            List<string> remoteLines = SplitLines(remote);
            List<string> localLines = SplitLines(local);
            long cellCount = (long)remoteLines.Count * localLines.Count;
            if (cellCount > 250000L)
            {
                return CoarseDiff(remoteLines, localLines);
            }

            int[,] table = new int[remoteLines.Count + 1, localLines.Count + 1];
            for (int remoteIndex = remoteLines.Count - 1; remoteIndex >= 0; remoteIndex--)
            {
                for (int localIndex = localLines.Count - 1; localIndex >= 0; localIndex--)
                {
                    if (remoteLines[remoteIndex] == localLines[localIndex])
                    {
                        table[remoteIndex, localIndex] = table[remoteIndex + 1, localIndex + 1] + 1;
                    }
                    else
                    {
                        table[remoteIndex, localIndex] = Math.Max(
                            table[remoteIndex + 1, localIndex],
                            table[remoteIndex, localIndex + 1]);
                    }
                }
            }

            var result = new List<PublishConflictDiffLine>();
            int remotePosition = 0;
            int localPosition = 0;
            while (remotePosition < remoteLines.Count && localPosition < localLines.Count)
            {
                if (remoteLines[remotePosition] == localLines[localPosition])
                {
                    Append(result, PublishConflictDiffLineKind.Same, remoteLines[remotePosition]);
                    remotePosition++;
                    localPosition++;
                }
                else if (table[remotePosition + 1, localPosition]
                    >= table[remotePosition, localPosition + 1])
                {
                    Append(result, PublishConflictDiffLineKind.Remote, remoteLines[remotePosition]);
                    remotePosition++;
                }
                else
                {
                    Append(result, PublishConflictDiffLineKind.Local, localLines[localPosition]);
                    localPosition++;
                }
            }

            while (remotePosition < remoteLines.Count)
            {
                Append(result, PublishConflictDiffLineKind.Remote, remoteLines[remotePosition]);
                remotePosition++;
            }

            while (localPosition < localLines.Count)
            {
                Append(result, PublishConflictDiffLineKind.Local, localLines[localPosition]);
                localPosition++;
            }

            return result;
        }

        private static IReadOnlyList<PublishConflictDiffLine> CoarseDiff(
            IReadOnlyList<string> remoteLines,
            IReadOnlyList<string> localLines)
        {
            var result = new List<PublishConflictDiffLine>(remoteLines.Count + localLines.Count);
            for (int index = 0; index < remoteLines.Count; index++)
            {
                Append(result, PublishConflictDiffLineKind.Remote, remoteLines[index]);
            }

            for (int index = 0; index < localLines.Count; index++)
            {
                Append(result, PublishConflictDiffLineKind.Local, localLines[index]);
            }

            return result;
        }

        private static void Append(
            List<PublishConflictDiffLine> result,
            PublishConflictDiffLineKind kind,
            string text)
        {
            result.Add(new PublishConflictDiffLine(result.Count, kind, text));
        }

        private static List<string> SplitLines(string value)
        {
            var result = new List<string>();
            int start = 0;
            for (int index = 0; index < value.Length; index++)
            {
                if (!IsNewline(value[index]))
                {
                    continue;
                }

                result.Add(value.Substring(start, index - start));
                start = index + 1;
            }

            result.Add(value.Substring(start));
            return result;
        }

        private static bool IsNewline(char value)
        {
            return value == '\u000A'
                || value == '\u000B'
                || value == '\u000C'
                || value == '\u000D'
                || value == '\u0085'
                || value == '\u2028'
                || value == '\u2029';
        }
    }
}

File: Shared/RepoPressCoreContracts/swift/Sources/RepoPressCore/PublishConflictDiffBuilder.swift
import Foundation

public struct PublishConflictDiffLine: Codable, Identifiable, Hashable, Sendable {
  public enum Kind: String, Codable, Hashable, Sendable {
    case same
    case remote
    case local
  }

  public var id: Int
  public var kind: Kind
  public var text: String

  public init(id: Int, kind: Kind, text: String) {
    self.id = id
    self.kind = kind
    self.text = text
  }

  public var marker: String {
    switch kind {
    case .same:
      return " "
    case .remote:
      return "-"
    case .local:
      return "+"
    }
  }
}

public struct PublishConflictDiffBuilder: Sendable {
  public init() {}

  public func diff(remote: String, local: String) -> [PublishConflictDiffLine] {
    let remoteLines = remote.components(separatedBy: .newlines)
    let localLines = local.components(separatedBy: .newlines)

    guard remoteLines.count * localLines.count <= 250_000 else {
      return coarseDiff(remoteLines: remoteLines, localLines: localLines)
    }

    var table = Array(
      repeating: Array(repeating: 0, count: localLines.count + 1),
      count: remoteLines.count + 1
    )

    if !remoteLines.isEmpty, !localLines.isEmpty {
      for remoteIndex in stride(from: remoteLines.count - 1, through: 0, by: -1) {
        for localIndex in stride(from: localLines.count - 1, through: 0, by: -1) {
          if remoteLines[remoteIndex] == localLines[localIndex] {
            table[remoteIndex][localIndex] = table[remoteIndex + 1][localIndex + 1] + 1
          } else {
            table[remoteIndex][localIndex] = max(
              table[remoteIndex + 1][localIndex],
              table[remoteIndex][localIndex + 1]
            )
          }
        }
      }
    }

    var result: [PublishConflictDiffLine] = []
    var remoteIndex = 0
    var localIndex = 0

    while remoteIndex < remoteLines.count, localIndex < localLines.count {
      if remoteLines[remoteIndex] == localLines[localIndex] {
        append(.same, remoteLines[remoteIndex], to: &result)
        remoteIndex += 1
        localIndex += 1
      } else if table[remoteIndex + 1][localIndex] >= table[remoteIndex][localIndex + 1] {
        append(.remote, remoteLines[remoteIndex], to: &result)
        remoteIndex += 1
      } else {
        append(.local, localLines[localIndex], to: &result)
        localIndex += 1
      }
    }

    while remoteIndex < remoteLines.count {
      append(.remote, remoteLines[remoteIndex], to: &result)
      remoteIndex += 1
    }

    while localIndex < localLines.count {
      append(.local, localLines[localIndex], to: &result)
      localIndex += 1
    }

    return result
  }

  private func coarseDiff(
    remoteLines: [String],
    localLines: [String]
  ) -> [PublishConflictDiffLine] {
    var result: [PublishConflictDiffLine] = []
    for line in remoteLines {
      append(.remote, line, to: &result)
    }
    for line in localLines {
      append(.local, line, to: &result)
    }
    return result
  }

  private func append(
    _ kind: PublishConflictDiffLine.Kind,
    _ text: String,
    to result: inout [PublishConflictDiffLine]
  ) {
    result.append(PublishConflictDiffLine(id: result.count, kind: kind, text: text))
  }
}
